test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs

```diff
@@ -1276,28 +1276,194 @@ public void GetLink_TwoOptionalParametersAfterDefault_LastValueFromAmbientValues
             Assert.Equal("/a", link);
         }
 
-        private RouteValuesBasedEndpointFinderContext CreateRouteValuesContext(object suppliedValues, object ambientValues = null)
+        public static TheoryData<object, object, object> DoesNotDiscardAmbientValuesData
         {
-            var context = new RouteValuesBasedEndpointFinderContext();
-            context.ExplicitValues = new RouteValueDictionary(suppliedValues);
-            context.AmbientValues = new RouteValueDictionary(ambientValues);
-            return context;
+            get
+            {
+                // - ambient values
+                // - explicit values
+                // - target endpoint required values
+                return new TheoryData<object, object, object>
+                {
+                    {
+                        new { controller = "Products", action = "Edit", id = 10 },
+                        new { controller = "Products", action = "Edit" },
+                        new { controller = "Products", action = "Edit" }
+                    },
+                    {
+                        new { controller = "ProDUcts", action = "EDit", id = 10 },
+                        new { controller = "ProDUcts", action = "EDit" },
+                        new { controller = "Products", action = "Edit" }
+                    },
+                    {
+                        new { area = "Admin", controller = "Products", action = "Edit", id = 10 },
+                        new { area = "Admin", controller = "Products", action = "Edit" },
+                        new { area = "Admin", controller = "Products", action = "Edit" }
+                    },
+                    {
+                        new { controller = "Products", action = "Edit", id = 10 },
+                        new { controller = "Products", action = "Edit" },
+                        new { area = (string)null, controller = "Products", action = "Edit" }
+                    },
```

Comment:
It would be good to include a case with pages too, for instance linking from a page to a controller and the reverse.